src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,433 @@
+}
+
+#if 0
+int ngx_http_lua_ffi_sem_get_value(ngx_http_lua_semaphore_t *sem,char **err,size_t* errlen)
+{
+    if (sem == NULL )
+    {
+        return 
+    }
+
+    return sem->count - sem->wait_count;
+}
+#endif
+
+
+int
+ngx_http_lua_ffi_sem_post(ngx_http_lua_semaphore_t *sem, int n, char **errstr)
+{
+    if (sem == NULL) {
+        *errstr = "semaphore is null";
+        return NGX_ERROR;
+    }
+
+    sem->count += n;
+
+    dd("ngx_http_lua_ffi_sem_post semaphore %p value %d", sem, sem->count);
+
+    if (!sem->in_post_event && !ngx_queue_empty(&sem->wait_queue)) {
+
+        sem->sem_event.handler = ngx_http_lua_ffi_semaphore_handler;
+        sem->sem_event.data = sem;
+        sem->sem_event.log = sem->log;
+
+        sem->in_post_event = 1;
+        ngx_post_event((&sem->sem_event), &ngx_posted_events);
+    }
+
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_ffi_sem_wait(ngx_http_request_t *r, ngx_http_lua_semaphore_t *sem,
+                          int time,u_char *errstr, size_t *errlen)
+{
+    ngx_http_lua_ctx_t           *ctx;
+    ngx_http_lua_co_ctx_t        *wait_co_ctx;
+    ngx_int_t                    rc;
+
+    if (r == NULL) {
+
+        *errlen = ngx_snprintf(errstr, *errlen,
+                               "request is null") - errstr;
+         return NGX_ERROR;
+    }
+
+    dd("ngx_http_lua_ffi_sem_wait semaphore: %p"
+       "value: %d in_post_event: %d", sem, sem->count, sem->in_post_event);
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        *errlen = ngx_snprintf(errstr, *errlen,
+                               "ctx is null") - errstr;
+        return NGX_ERROR;
+    }
+
+    rc = ngx_http_lua_ffi_check_context(ctx, (NGX_HTTP_LUA_CONTEXT_REWRITE
+                                        | NGX_HTTP_LUA_CONTEXT_ACCESS
+                                        | NGX_HTTP_LUA_CONTEXT_CONTENT
+                                        | NGX_HTTP_LUA_CONTEXT_TIMER),
+                                        errstr, errlen);
+
+    if (rc == NGX_DECLINED) {
+        return NGX_ERROR;
+    }
+
+    if (!sem->in_post_event && sem->count > 0) {
+         sem->count--;
+         return NGX_OK;
+
+    } else if (time == 0) {
+        return NGX_DECLINED;
+
+    } else {
+         sem->wait_count++;
+         wait_co_ctx = ctx->cur_co_ctx;
+         wait_co_ctx->sleep.handler = ngx_http_lua_sem_timeout_handler;
+         wait_co_ctx->sleep.data = ctx->cur_co_ctx;
+         wait_co_ctx->sleep.log = r->connection->log;
+
+         ngx_add_timer(&wait_co_ctx->sleep, (ngx_msec_t) time * 1000);
+
+         dd("ngx_http_lua_ffi_sem_wait add timer coctx:%p time:%d(s)",
+             wait_co_ctx, time);
+
+         ngx_queue_insert_tail(&sem->wait_queue, &wait_co_ctx->sem_wait_queue);
+         wait_co_ctx->data = sem;
+         wait_co_ctx->cleanup = ngx_http_lua_ffi_semaphore_cleanup;
+
+         return NGX_AGAIN;
+    }
+}
+
+
+static void
+ngx_http_lua_ffi_semaphore_cleanup(void *data)
```

Comment:
Hmm, some comments:

1. I suggest using the `ngx_http_lua_ffi_` prefix for public C functions that are part of the ABI.
2. I see you are using both "sem" and "semaphore" in your C function names, which is inconsistent. Please be consistent :)

Thanks!